sympy/solvers/bivariate.py

```diff
@@ -176,22 +183,83 @@ def _lambert(eq, x):
     p, den = den.as_coeff_Mul()
     e = exp(num/den)
     t = Dummy('t')
-    args = [d/(a*b)*t for t in roots(t**p - e, t).keys()]
 
+    if p == 1:
+        t = e
+        args = [d/(a*b)*t]
+    elif domain.is_real:
+        ind_ls = [d/(a*b)*t for t in roots(t**p - e, t).keys()]
+        args = []
+        j = -1
+        for i in ind_ls:
+            j += 1
+            if not isinstance(i,int):
+                if not i.has(I):
+                    args.append(ind_ls[j])
+            elif isinstance(i,int):
+                args.append(ind_ls[j])
+    else:
+        args = [d/(a*b)*t for t in roots(t**p - e, t).keys()]
+    if len(args) == 0:
+        return S.EmptySet
     # calculating solutions from args
+
+    integer = Symbol('integer',integer=True)
     for arg in args:
-        for k in lambert_real_branches:
-            w = LambertW(arg, k)
-            if k and not w.is_real:
-                continue
-            rhs = -c/b + (a/d)*w
-
-            for xu in xusolns:
-                sol.append(xu.subs(u, rhs))
+        if not domain.is_subset(S.Reals):
+            rhs = -c/b + (a/d)*LambertW(arg,integer)
+
+        if not len(list(eq.atoms(Symbol,Dummy))) >1:
+            if domain.is_subset(S.Reals):
+                if -1/E <= re(arg) < 0 and LambertW(arg).is_real:
+                    for k in lambert_real_branches:
+                        rhs = -c/b + (a/d)*LambertW(arg, k)
```

Comment:
This ode and the lines below are messy and repetitive which makes it hard to understand what the actual logic here is doing. That should be simplified.